ipynb/energy/SystemEnergy: Rework pstate modelling

valschneider opened the pull request:
This is fairly extensive rework of the SystemEnergy notebook.
The main additions are:

* Computation of "tare" (noise) system energy. With the previous
  implementation, we would determine that the energy measured
  at (n_cpu == 0) would be the cluster energy. However, this
  energy includes a lot of other components, even more so when
  the energy measurement point is far away from the clusters
  (from the standpoint of the electrical circuit).

  The proposed improvement is as follows:
  - Before doing measurement on cluster 'A', find another cluster 'B'.
  - Offline cluster A, online a single CPU noise_cpu of cluster B
  - Run benchmark on noise_cpu. Energy measurement = noise_energy
  - Online cluster A and run the tests.
    Run benchmark on noise_cpu during the tests
    Subtract noise_energy from test energy measurement

  While not flawless, this should greatly improve our cluster costs.

* Energy measurement filtering. Due to many different physical
  parameters, energy curves can be non-monotonic, which shouldn't be
  the case. As such, non-monotonic readings are expunged.

* Energy model monotonization. Even with model generation improvements,
  the resulting energy model can be non-monotonic and we then have to
  "massage" it into shape. This is now done inside the notebook itself
  by replacing non-monotonic values with a polynomial interpolation.
  This isn't perfect, as I've had cases where the interpolation
  itself wouldn't be monotonic (locally, within the scope of our
  energy values), but this can easily be spotted.

mdigiorgio commented on ipynb/energy/EnergyModel_SystemEnergy.ipynb:
```diff
@@ -452,404 +457,1647 @@
    "outputs": [
     {
      "data": {
-      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA7sAAAJqCAYAAADqj1fIAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzs3Xl4VdW9//H3CkPICAkgBNAglUFAQFEUUWRUrrciEioF\nlLGi9mrrD22lWBmEetVylQvSihRLQ0EqhVagV4uIgIAgKmMQIkEDhohCCJkgQLJ+f+yTw0lyMjHk\n7CSf1/OcJ9l7rbXP2vt80XzPWmtvY61FREREREREpDoJCnQHRERERERERC43JbsiIiIiIiJS7SjZ\nFRERERERkWpHya6IiIiIiIhUO0p2RUREREREpNpRsisiIiIiIiLVjpJdEZEAMsaMMsb ...
```
`plot_data()` maybe?

mdigiorgio replied:
s/Im/In/

mdigiorgio replied:
Could you add documentation for `metric`?

valschneider commented on ipynb/energy/EnergyModel_SystemEnergy.ipynb:
```diff
@@ -79,11 +71,12 @@
     "import re\n",
     "import trappy\n",
     "\n",
-    "from collections import namedtuple, OrderedDict\n",
```
I think the answer would be "no real reason in particular". I wanted the `clusters` dict to be populated automatically instead of manually, so I sketched a quick piece of code without minding the previous `namedtuple`. I could change that back to `namedtuple`, but I don't find the dict accessors to be that much of a nuisance (personal opinion here).

mdigiorgio replied:
Alright, I was just wondering. Automatic filling seems a good reason to switch :)

valschneider commented on ipynb/energy/EnergyModel_SystemEnergy.ipynb:
```diff
@@ -251,119 +240,126 @@
+    "clusters = {}\n",
+    "\n",
+    "for cpu, cluster in enumerate(target.core_clusters):\n",
+    "    if clusters.get(cluster, None) is None:\n",
```
Looking at this again it does look convoluted... Thanks.

valschneider commented on ipynb/energy/EnergyModel_SystemEnergy.ipynb:
```diff
@@ -452,404 +457,1647 @@
+    "            \n",
+    "            # Per-iteration delay\n",
+    "            if self.delay > 0:\n",
+    "                sleep(self.delay)\n",
```
It is, but I think it's better to ask the question: "do I need to sleep ?" rather than do "give the sleep function whatever argument I have and it'll take care of it for me". That's me being pedantic though.

valschneider commented:
Thanks @mdigiorgio for the comments, I'll do some more cleanup & doc writing and then update this PR.

valschneider commented:
Did some cleanup, added some comments/documentation, and also changed the iteration delay in a temperature lookup - we wait for temperature to drop down below a certain threshold (determined when the tests are started) before running the next test iteration.
